runtime/cudaq/platform/default/rest_server/RestServerMain.cpp

```diff
@@ -0,0 +1,221 @@
+  sourceMgr.AddNewSourceBuffer(llvm::MemoryBuffer::getMemBufferCopy(irString),
+                               llvm::SMLoc());
+  auto module = parseSourceFile<ModuleOp>(sourceMgr, contextPtr.get());
+  auto engine = jitCode(*module, passes);
+  const std::string entryPointFunc =
+      std::string(cudaq::runtime::cudaqGenPrefixName) + entryPointFn;
+  auto fnPtr = llvm::cantFail(engine->lookup(entryPointFunc));
+  if (!fnPtr)
+    throw std::runtime_error("Failed to get entry function");
+
+  auto fn = reinterpret_cast<void (*)()>(fnPtr);
+  // Invoke the kernel
+  fn();
+}
+
+void *loadNvqirSimLib(const std::string &simulatorName) {
+  const std::filesystem::path cudaqLibPath{cudaq::getCUDAQLibraryPath()};
+#if defined(__APPLE__) && defined(__MACH__)
+  const auto libSuffix = "dylib";
+#else
+  const auto libSuffix = "so";
+#endif
+  const auto simLibPath =
+      cudaqLibPath.parent_path() /
+      fmt::format("libnvqir-{}.{}", simulatorName, libSuffix);
+  cudaq::info("Request simulator {} at {}", simulatorName, simLibPath.c_str());
+  void *simLibHandle = dlopen(simLibPath.c_str(), RTLD_GLOBAL | RTLD_NOW);
+  if (!simLibHandle) {
+    char *error_msg = dlerror();
+    throw std::runtime_error(fmt::format(
+        "Failed to open simulator backend library: {}.",
+        error_msg ? std::string(error_msg) : std::string("Unknown error")));
+  }
+  auto *sim = cudaq::getUniquePluginInstance<nvqir::CircuitSimulator>(
+      std::string("getCircuitSimulator"), simLibPath.c_str());
+  __nvqir__setCircuitSimulator(sim);
+
+  return simLibHandle;
+}
+
+json processRequest(const std::string &reqBody) {
+  auto requestJson = json::parse(reqBody);
+  cudaq::RestRequest request(requestJson);
+  void *handle = loadNvqirSimLib(request.simulator);
+  if (request.seed != 0)
+    cudaq::set_random_seed(request.seed);
+  auto mlirContext = cudaq::initializeMLIR();
+  auto &platform = cudaq::get_platform();
+  platform.set_exec_ctx(&request.executionContext);
+  std::vector<char> decodedCodeIr;
+  if (llvm::decodeBase64(request.code, decodedCodeIr))
+    throw std::runtime_error("Failed to decode input IR");
+
+  std::string_view codeStr(decodedCodeIr.data(), decodedCodeIr.size());
+  if (request.format == cudaq::CodeFormat::LLVM)
+    cudaq::invokeWrappedKernel(codeStr, request.entryPoint, request.args);
+  else
+    invokeMlirKernel(mlirContext, codeStr, request.passes, request.entryPoint);
+  platform.reset_exec_ctx();
+  json resultContextJs = request.executionContext;
+  dlclose(handle);
+  return resultContextJs;
+}
+} // namespace
+
+int main(int argc, char **argv) {
+  constexpr int DEFAULT_PORT = 3030;
+  const int port = [&]() {
+    for (int i = 1; i < argc - 1; ++i)
+      if (std::string(argv[i]) == "-p" || std::string(argv[i]) == "-port" ||
+          std::string(argv[i]) == "--port")
+        return atoi(argv[i + 1]);
+
+    return DEFAULT_PORT;
+  }();
+  cudaq::mpi::initialize();
+  cudaq::RestServer server(port);
```

Comment:
We might consider how to extract the "execution logic" out here into a separate extensible interface - enabling this runtime server daemon to be deployed to partner QPUs and allowing partners to control how our IR gets translated to control system handling.  (comment / question from today's walk through)